samples/file-encryptor/enclave/openssl_src/keys.cpp

```diff
@@ -0,0 +1,428 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openssl/aes.h>
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+#include <string.h>
+
+#include "common/encryptor.h"
+#include "common/trace.h"
+
+#define ENCRYPT_OPERATION true
+#define DECRYPT_OPERATION false
+
+// Compute the sha256 hash of given data.
+int ecall_dispatcher::Sha256(
+    const uint8_t* data,
+    size_t data_size,
+    uint8_t sha256[32])
+{
+    int ret = 1;
+    unsigned int sha_length_in_bytes = 32;
+    EVP_MD_CTX* mdctx;
+
+    if ((mdctx = EVP_MD_CTX_new()) == nullptr)
+    {
+        TRACE_ENCLAVE("EVP_MD_CTX_new failed");
+        goto exit;
+    }
+
+    if (!(ret = EVP_DigestInit_ex(mdctx, EVP_sha256(), nullptr)))
+    {
+        TRACE_ENCLAVE(
+            "EVP_DigestInit_ex SHA-256 failed with returncode %d", ret);
+        goto exit;
+    }
+
+    if (!(ret = EVP_DigestUpdate(mdctx, data, data_size)))
+    {
+        TRACE_ENCLAVE("EVP_DigestUpdate failed with returncode %d", ret);
+        goto exit;
+    }
+
+    if (!(ret = EVP_DigestFinal_ex(mdctx, sha256, &sha_length_in_bytes)))
+    {
+        TRACE_ENCLAVE("EVP_DigestFinal_ex failed with return code %d", ret);
+        goto exit;
+    }
+
+    ret = 0;
+exit:
+    EVP_MD_CTX_free(mdctx);
+    return ret;
+}
+
+// This routine uses the openssl library to derive an AES key from the input
+// password and produce a password based key.
+int ecall_dispatcher::generate_password_key(
+    const char* password,
+    unsigned char* salt,
+    unsigned char* key,
+    unsigned int key_size)
+{
+    int ret = 1;
+
+    if (PKCS5_PBKDF2_HMAC(
+            (const char*)password,              // passwd
+            (int)strlen((const char*)password), // passwd length
+            salt,                               // salt
+            sizeof(salt),                       // salt length
```

Comment:
BUG: this is incorrect since salt is a pointer, so sizeof(salt) would be the pointer size, not the size of the memory pointed to.  Here you can use SALT_SIZE_IN_BYTES or make the caller pass in salt_size.
```suggestion
            SALT_SIZE_IN_BYTES,
```